File: PathoFroh/src/main/java/com/patho/main/service/UserService.java
package com.patho.main.service;

import java.util.List;
import java.util.Optional;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

import com.patho.main.common.ContactRole;
import com.patho.main.model.Contact;
import com.patho.main.model.Person;
import com.patho.main.model.Physician;
import com.patho.main.model.user.HistoGroup;
import com.patho.main.model.user.HistoSettings;
import com.patho.main.model.user.HistoUser;
import com.patho.main.repository.GroupRepository;
import com.patho.main.repository.LDAPRepository;
import com.patho.main.repository.UserRepository;

@Service
@Transactional
public class UserService extends AbstractService {

	private final Logger logger = LoggerFactory.getLogger(this.getClass());

	@Autowired
	private UserRepository userRepository;

	@Autowired
	private UserService userService;

	@Autowired
	private PhysicianService physicianService;

	@Autowired
	private GroupRepository groupRepository;

	@Autowired
	private LDAPRepository ldapRepository;

	/**
	 * Creats a new user, loads data from ldap and creates the user with the guest
	 * group
	 * 
	 * @param uid
	 * @return
	 */
	public Optional<HistoUser> createUser(String uid) {
		Optional<Physician> oPhyisican = ldapRepository.findByUid(uid);
		Optional<HistoGroup> group = groupRepository.findOptionalById(HistoGroup.GROUP_GUEST_ID);

		if (oPhyisican.isPresent() && group.isPresent()) {
			HistoUser user = addOrMergeUser(oPhyisican.get());

			if (user != null) {
				user = updateGroupOfUser(user, group.get());
				return Optional.ofNullable(user);
			}
		}

		return Optional.empty();
	}

	/**
	 * Adds a histouser and updates the passed physician.
	 * 
	 * @param physician
	 * @return
	 */
	public HistoUser addOrMergeUser(Physician physician) {
		return addOrMergeUser(physician, true);
	}

	/**
	 * Creates an new user, if a physician with the given username exist they will
	 * be merged.
	 * 
	 * @param physician
	 * @return
	 */
	public HistoUser addOrMergeUser(Physician physician, boolean checkPhysican) {

		if (physician == null) {
			return null;
		}

		// checking if histouser exsists
		Optional<HistoUser> histoUser = userRepository.findOptionalByPhysicianUid(physician.getUid());

		if (!histoUser.isPresent()) {
			logger.info("No User found, creating new HistoUser " + physician.getPerson().getFullName());

			if (checkPhysican)
				physician = physicianService.addOrMergePhysician(physician);

			HistoUser newHistoUser = new HistoUser(physician);

			if (physician.getAssociatedRoles().size() == 0)
				physician.addAssociateRole(ContactRole.NONE);

			// saving or updating physician, also updating organizations
			physician = physicianService.addOrMergePhysician(physician);

			newHistoUser.setPhysician(physician);

			return userRepository.save(newHistoUser, resourceBundle.get("log.user.created", histoUser));
		} else {
			histoUser.get().setPhysician(physicianService.mergePhysicians(physician, histoUser.get().getPhysician()));
			return userRepository.save(histoUser.get(), resourceBundle.get("log.user.update", histoUser));
		}
	}

	/**
	 * Updates the users data with data from ldap
	 * 
	 * @param user
	 */
	public HistoUser updateUserWithLdapData(HistoUser user) {
		user.setPhysician(physicianService.updatePhysicianWithLdapData(user.getPhysician()));
		return user;
	}

	/**
	 * Save user
	 * 
	 * @param user
	 */
	public void saveUser(HistoUser user) {
		if (user.getPhysician().hasNoAssociateRole())
			user.getPhysician().addAssociateRole(ContactRole.OTHER_PHYSICIAN);

		user.setPhysician(physicianService.addOrMergePhysician(user.getPhysician()));
		userRepository.save(user, resourceBundle.get("log.user.update", user));
	}

	@Transactional(propagation = Propagation.NEVER)
	public boolean deleteOrDisable(HistoUser user, boolean deletePhysician) {
		Physician physician = user.getPhysician();

		try {
			// setting dummy for deleting user
			user.setPhysician(new Physician(new Person(new Contact())));

			logger.debug("Deleting user");
			// error is only fired by the delete command
			user = userRepository.save(user);
			userRepository.delete(user,
					resourceBundle.get("llog.user.deleted", user.getPhysician().getPerson().getFullName()));

			if (deletePhysician) {
				logger.debug("deleting physician");
				physicianService.deleteOrArchive(physician);
			} else
				logger.debug("keeping physician");

			return true;
		} catch (Exception e) {
			// reverting user disable
			user.setPhysician(physician);
			logger.debug("deleting not possible, disable");
			diableUser(user);

			// only archive
			if (deletePhysician) {
				logger.debug("physician is not in use, archive");
				physicianService.archive(physician, true);
			} else
				logger.debug("physician is in use, do nothing");

			return false;
		}
	}

	/**
	 * Disable User via group
	 * 
	 * @param user
	 */
	public HistoUser diableUser(HistoUser user) {
		// do not load settings is done by the updateGroupOfUser methode
		Optional<HistoGroup> oGroup = groupRepository.findOptionalById(HistoGroup.GROUP_DISABLED, false);

		if (oGroup.isPresent())
			return updateGroupOfUser(user, oGroup.get());
		return user;

	}

	/**
	 * Archive User
	 * 
	 * @param user
	 * @param archive
	 */
	public void archiveUser(HistoUser user, boolean archive) {
		user.setArchived(archive);
		userRepository.save(user, resourceBundle.get(archive ? "log.user.archived" : "log.user.dearchived", user));
	}

	/**
	 * Updates the user group settings for a changed groups
	 * 
	 * @param user
	 * @param group
	 */
	public void updateGroupOfUser(HistoGroup group) {
		// updating user settings
		List<HistoUser> usersOfGroup = userRepository.findByGroupId(group.getId());
		usersOfGroup.forEach(p -> updateGroupOfUser(p, group));
	}

	/**
	 * Updates the group of an user
	 * 
	 * @param user
	 * @param group
	 * @return
	 */
	public HistoUser updateGroupOfUser(HistoUser user, HistoGroup group) {
		return updateGroupOfUser(user, group, true);
	}

	/**
	 * Updates the group of an user
	 * 
	 * @param user
	 * @param group
	 * @return
	 */
	public HistoUser updateGroupOfUser(HistoUser user, long groupID, boolean save) {
		Optional<HistoGroup> group = groupRepository.findOptionalById(groupID, true);
		return updateGroupOfUser(user, group.orElse(new HistoGroup(new HistoSettings())), save);
	}

	/**
	 * Setting the group of the user
	 * 
	 * @param user
	 * @param group
	 */
	public HistoUser updateGroupOfUser(HistoUser user, HistoGroup group, boolean save) {
		user.setGroup(group);

		if (user.getSettings() == null)
			user.setSettings(new HistoSettings());

		GroupService.copyGroupSettings(user, group, false);

		logger.debug("Role of user " + user.getUsername() + " to " + user.getGroup().toString());

		if (save)
			return userRepository.save(user, resourceBundle.get("log.user.role.changed", user.getGroup()));

		return user;
	}

}

File: PathoFroh/src/main/kotlin/com/patho/main/service/UserService.kt
package com.patho.main.service

import com.patho.main.common.ContactRole
import com.patho.main.model.Physician
import com.patho.main.model.person.Contact
import com.patho.main.model.person.Person
import com.patho.main.model.user.HistoGroup
import com.patho.main.model.user.HistoUser
import com.patho.main.repository.GroupRepository
import com.patho.main.repository.LDAPRepository
import com.patho.main.repository.UserRepository
import com.patho.main.util.user.HistoGroupNotFoundException
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Propagation
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service()
open class UserService @Autowired constructor(
        private val ldapRepository: LDAPRepository,
        private val groupRepository: GroupRepository,
        private val userRepository: UserRepository,
        private val physicianService: PhysicianService) : AbstractService() {

    /**
     * Checks if the session is associated with a user.
     *
     * @return
     */
    val isCurrentUserAvailable: Boolean
        get() = SecurityContextHolder.getContext().authentication.principal is HistoUser

    /**
     * Returns the current user.
     *
     * @return
     */
    val currentUser: HistoUser
        get() = SecurityContextHolder.getContext().authentication.principal as HistoUser

    /**
     * Creates a new user, loads data from ldap and creates the user with the guest
     * group
     *
     * @param uid
     * @return
     */
    @Transactional
    open fun createUser(uid: String): Optional<HistoUser> {
        val physician = ldapRepository.findByUid(uid)
        val group = groupRepository.findOptionalById(HistoGroup.GROUP_GUEST_ID)

        if (physician.isPresent && group.isPresent) {
            var user: HistoUser? = addOrMergeUser(physician.get())

            if (user != null) {
                user = updateGroupOfUser(user, group.get())
                return Optional.ofNullable(user)
            }
        }

        return Optional.empty()
    }

    /**
     * Adds a histouser and updates the passed physician.
     */
    @Transactional
    open fun addOrMergeUser(physician: Physician): HistoUser? {
        return addOrMergeUser(physician, true)
    }

    /**
     * Creates an new user, if a physician with the given username exist they will
     * be merged.
     */
    @Transactional
    open fun addOrMergeUser(physician: Physician, checkPhysician: Boolean): HistoUser {

        // checking if histouser exsists
        val histoUser = userRepository.findOptionalByPhysicianUid(physician.uid)

        if (!histoUser.isPresent) {
            var tmpPhysician = physician
            logger.info("No User found, creating new HistoUser " + tmpPhysician.person.getFullName())

            if (checkPhysician)
                tmpPhysician = physicianService.addOrMergePhysician(tmpPhysician)

            val newHistoUser = HistoUser(tmpPhysician)

            if (tmpPhysician.hasNoAssociateRole())
                tmpPhysician.addAssociatedRole(ContactRole.NONE)

            // saving or updating physician, also updating organizations
            tmpPhysician = physicianService.addOrMergePhysician(tmpPhysician)

            newHistoUser.physician = tmpPhysician

            return userRepository.save(newHistoUser, resourceBundle.get("log.user.created", histoUser))
        } else {
            histoUser.get().physician = physicianService.mergePhysicians(physician, histoUser.get().physician)
            return userRepository.save(histoUser.get(), resourceBundle.get("log.user.update", histoUser))
        }
    }

    /**
     * Updates the users data with data from ldap
     */
    @Transactional
    open fun updateUserWithLdapData(user: HistoUser): HistoUser {
        user.physician = physicianService.updatePhysicianWithLdapData(user.physician)
        return user
    }

    /**
     * Saves the user
     */
    @Transactional
    open fun saveUser(user: HistoUser) {
        if (user.physician.hasNoAssociateRole())
            user.physician.addAssociatedRole(ContactRole.OTHER_PHYSICIAN)
        user.physician = physicianService.addOrMergePhysician(user.physician)
        userRepository.save(user, resourceBundle.get("log.user.update", user))
    }

    /**
     * This method tries to delete a user, if this is not possible, the user will be archived
     */
    @Transactional(propagation = Propagation.NEVER)
    open fun deleteOrDisable(user: HistoUser, deletePhysician: Boolean): Boolean {
        var user = user
        val physician = user.physician

        try {
            // setting dummy for deleting user, this is for preserving the original physician
            user.physician = Physician(Person(Contact()))

            logger.debug("Deleting user")
            // error is only fired by the delete command
            user = userRepository.save(user)

            userRepository.delete(user,
                    resourceBundle.get("log.user.deleted", user.physician.person.getFullName()))

            if (deletePhysician) {
                logger.debug("Deleting physician")
                physicianService.deleteOrArchive(physician)
            } else
                logger.debug("Keeping physician")

            return true
        } catch (e: Exception) {
            // reverting user disable
            user.physician = physician
            logger.debug("deleting not possible, disable")
            disableUser(user)

            // only archive
            if (deletePhysician) {
                logger.debug("Physician is not in use, arching...")
                physicianService.archive(physician, true)
            } else
                logger.debug("Physician is in use, do nothing")

            return false
        }

    }

    /**
     * Disables a user via group settings
     *
     * @param user
     */
    @Transactional
    open fun disableUser(user: HistoUser): HistoUser {
        // do not load settings is done by the updateGroupOfUser methode
        val group = groupRepository.findOptionalById(HistoGroup.GROUP_DISABLED, false)

        if (!group.isPresent)
            throw HistoGroupNotFoundException()

        return updateGroupOfUser(user, group.get())

    }

    /**
     * Archives a user
     *
     * @param user
     * @param archive
     */
    @Transactional
    open fun archiveUser(user: HistoUser, archive: Boolean) {
        user.archived = archive
        userRepository.save(user, resourceBundle.get(if (archive) "log.user.archived" else "log.user.dearchived", user))
    }

    /**
     * Updates the user's group settings for a changed groups
     *
     * @param user
     * @param group
     */
    @Transactional
    open fun updateGroupOfUser(group: HistoGroup) {
        // updating user settings
        val usersOfGroup = userRepository.findByGroupId(group.id)
        usersOfGroup.forEach { p -> updateGroupOfUser(p, group) }
    }

    /**
     * Updates the group of an user
     *
     * @param user
     * @param group
     * @return
     */
    @Transactional
    open fun updateGroupOfUser(user: HistoUser, group: HistoGroup): HistoUser {
        return updateGroupOfUser(user, group, true)
    }

    /**
     * Updates the group of an user
     *
     * @param user
     * @param group
     * @return
     */
    @Transactional
    open fun updateGroupOfUser(user: HistoUser, groupID: Long, save: Boolean): HistoUser {
        val group = groupRepository.findOptionalById(groupID, true)
        if (!group.isPresent)
            throw  HistoGroupNotFoundException()
        return updateGroupOfUser(user, group.get(), save)
    }

    /**
     * Setting the group of the user
     */
    @Transactional
    open fun updateGroupOfUser(user: HistoUser, group: HistoGroup, save: Boolean): HistoUser {
        user.group = group

        GroupService.copyGroupSettings(user, group, false)

        logger.debug("Role of user " + user.username + " to " + user.group.toString())

        return if (save) userRepository.save(user, resourceBundle.get("log.user.roleChanged", user.group)) else user

    }
}